TritonInferenceServer/utilities/utilization.py

```diff
@@ -0,0 +1,229 @@
+# Copyright 2018-2020 MobiledgeX, Inc. All rights and licenses reserved.
+# MobiledgeX, Inc. 156 2nd Street #408, San Francisco, CA 94105
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+import os
+import subprocess
+import struct
+import json
+import time
+import logging
+import datetime
+import time
+import psutil
+import logging
+import argparse
+from subprocess import Popen, PIPE
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+formatter = logging.Formatter('%(asctime)s - %(threadName)s - %(levelname)s - %(message)s')
+ch = logging.StreamHandler()
+ch.setLevel(logging.INFO)
+ch.setFormatter(formatter)
+logger.addHandler(ch)
+fh = logging.FileHandler(__name__+'.log')
+fh.setLevel(logging.DEBUG)
+fh.setFormatter(formatter)
+
+gpu_supported = True
+
+def usage_gpu():
+    """
+    Execute the "nvidia-smi" command and parse the output to get the GPU usage.
+    """
+    global gpu_supported
+    if not gpu_supported:
+        return {}
+
+    nvidia_smi = "nvidia-smi"
+    try:
+        p = Popen([nvidia_smi, "-q", "-d", "UTILIZATION"], stdout=PIPE)
+    except FileNotFoundError:
+        logger.error("%s not found. Check if your Nvidia driver install is correct. GPU usage not supported." %nvidia_smi)
+        return {} # Empty dict
+
+    stdout, stderror = p.communicate()
+    output = stdout.decode('UTF-8')
+    # Split on line break
+    lines = output.split(os.linesep)
+
+    index = 0
+    while index < len(lines):
+        line = lines[index]
+        if line.strip() == "Utilization":
+            line = lines[index+1]
+            vals = line.split()
+            gpu_util_snap = vals[2]
+            line = lines[index+2]
+            vals = line.split()
+            gpu_mem_util_snap = vals[2]
```

Comment:
Where these indices come from? Maybe put them in a variable to make it clear (eg. utilization_index or utilization_index_offset)? Or comment is fine too.

Reply:
I've been meaning to redo this since I found better options for the nvidia-smi command to get the exact data I want without having to parse a bunch of output. You have convinced me to go ahead and implement it.